File: py/structural/adapter.py
# Adapter wraps an incompatible object and exposes the interface a client expects.
# Modern Python note: duck typing reduces the need, but adapters are still valuable
# at API, library, network, and legacy-system boundaries.
class EuropeanSocket:
    def voltage(self):
        return 230


class PhoneCharger:
    def charge(self, usb_power_source):
        return f"charging at {usb_power_source.usb_voltage()}V"


class EuropeanToUsbAdapter:
    def __init__(self, socket):
        self.socket = socket

    def usb_voltage(self):
        wall_voltage = self.socket.voltage()
        return wall_voltage // 46


if __name__ == "__main__":
    charger = PhoneCharger()
    adapter = EuropeanToUsbAdapter(EuropeanSocket())
    print(charger.charge(adapter))

File: csharp/structural/adapter.cs
namespace Gof.ModernCSharp.Structural;

// Adapter wraps an incompatible object and exposes the interface a client expects.
// Modern C# note: still useful at package, HTTP, cloud SDK, and legacy boundaries.
public static class Adapter
{
    sealed class EuropeanSocket { public int Voltage() => 230; }
    interface IUsbPowerSource { int UsbVoltage(); }
    sealed class PhoneCharger { public string Charge(IUsbPowerSource source) => $"charging at {source.UsbVoltage()}V"; }

    sealed class EuropeanToUsbAdapter(EuropeanSocket socket) : IUsbPowerSource
    {
        public int UsbVoltage() => socket.Voltage() / 46;
    }

    public static string Run() => new PhoneCharger().Charge(new EuropeanToUsbAdapter(new EuropeanSocket()));
}
